File: roles/ngfw_lab/files/monitor.py
#!/usr/bin/env python3
"""Loopback-only, read-only live UI for existing runner artifacts. Never starts tests."""
import argparse
from collections import deque
from datetime import datetime, timezone
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
import math
from pathlib import Path
import re
import time
from urllib.parse import urlsplit

ASSETS = Path(__file__).with_name('monitor')


def finite(value):
    return value if type(value) in (int, float) and math.isfinite(value) else None


def queue_projection(value):
    value = value or {}
    allowed = ('size', 'enqueued', 'full', 'discarded.full', 'discarded.nf', 'maxqsize',
               'processed', 'failed', 'suspended', 'resumed', 'suspended.duration', 'submitted')
    return {'age_seconds': finite(value.get('age_seconds')), 'stale': value.get('stale') is True,
            'counters': {str(name)[:100]: {k: finite(row.get(k)) for k in allowed if k in row}
                         for name, row in (value.get('counters') or {}).items() if isinstance(row, dict)}}


def read_json(path, limit=2097152):
    if path.is_symlink():
        raise ValueError('symlink artifact rejected')
    with path.open('rb') as src:
        value = src.read(limit + 1)
    if len(value) > limit:
        raise ValueError('artifact exceeds display bound')
    return json.loads(value)


def tail_rows(path, limit=4194304, rows=360):
    if path.is_symlink():
        raise ValueError('symlink artifact rejected')
    with path.open('rb') as src:
        src.seek(0, 2)
        size = src.tell()
        src.seek(max(0, size - limit))
        data = src.read(limit)
    lines = data.splitlines()
    if size > limit:
        lines = lines[1:]
    output = deque(maxlen=rows)
    for line in lines:
        try:
            row = json.loads(line)
            if isinstance(row, dict):
                output.append(row)
        except (ValueError, UnicodeDecodeError):
            pass  # A partially appended final row is not a new measurement.
    return list(output)


def project_sample(row):
    observed = row.get('measurement') or {}
    guest, host = observed.get('guest') or {}, observed.get('host') or {}
    gd, hd = guest.get('derived') or {}, host.get('derived') or {}
    audit = guest.get('audit') or {}
    legacy = row.get('audit') or {}
    cpus = {name: finite(val.get('busy_pct')) for name, val in gd.get('cpu', {}).items() if name != 'cpu'}
    baseline = guest.get('busy_poll_baseline') or {}
    polling = [name for name in baseline.get('verified_cores', [])
               if isinstance(name, str) and re.fullmatch(r'cpu[0-9]{1,3}', name) and name in cpus]
    temperature = {name: finite(value) for name, value in (host.get('temperature_c') or {}).items()}
    disks = {name: {k: finite(v.get(k)) for k in ('await_ms', 'write_bytes_s', 'iops', 'avg_queue_depth')}
             for name, v in gd.get('disks', {}).items()}
    backlog, limit = audit.get('backlog', legacy.get('audit_backlog')), audit.get('backlog_limit', legacy.get('audit_backlog_limit'))
    return {'time': row.get('time'), 'phase': row.get('phase'), 'scenario': row.get('scenario'),
            'guest_cpu_pct': finite(gd.get('cpu', {}).get('cpu', {}).get('busy_pct', legacy.get('cpu_pct'))),
            'host_cpu_pct': finite(hd.get('cpu', {}).get('cpu', {}).get('busy_pct')),
            'guest_cpu_breakdown': {k: finite(gd.get('cpu', {}).get('cpu', {}).get(k + '_pct'))
                                    for k in ('system', 'iowait', 'irq', 'softirq', 'steal')},
            'processes': [{k: value.get(k) for k in ('comm', 'cpu_one_core_pct', 'state', 'wchan')}
                          for value in gd.get('processes', {}).values()],
            'cores': cpus, 'verified_polling_cores': polling,
            'temperature_c': temperature, 'disks': disks,
            'backlog': finite(backlog), 'backlog_pct': 100 * backlog / limit if finite(backlog) is not None and finite(limit) and limit > 0 else None,
            'lost': finite(audit.get('lost', legacy.get('audit_lost'))),
            'audit_wait_delta': finite(gd.get('audit_wait_delta')),
            'audit_bytes_s': finite(gd.get('log_bytes_per_second', legacy.get('log_bytes_per_second'))),
            'rotation': bool(gd.get('rotation')),
            'forwarder': queue_projection(guest.get('forwarder')),
            'collector': {'counters': [{'id': str(v.get('id', ''))[:100], 'counter': str(v.get('counter', ''))[:100],
                                        'value': finite(v.get('value'))}
                                       for v in (row.get('collector') or {}).get('counters', [])[:100]]},
            'endpoint_rx_bps': finite(row.get('endpoint_rx_bps')),
            'service_probe_ms': finite(row.get('service_probe_ms')),
            'collection_seconds': finite(row.get('collection_seconds')),
            'unavailable': [str(x)[:160] for x in (row.get('unavailable', []) +
                ['guest: ' + str(g) for g in guest.get('gaps', [])] +
                ['host: ' + str(g) for g in host.get('gaps', [])])][:32],
            'checks': {k: row.get(k) is True for k in ('dataplane_ok', 'management_ok', 'mngt_ok', 'vms_ok')}}


class Store:
    def __init__(self, root):
        self.root = Path(root).resolve()

    def directory(self, name):
        if not re.fullmatch(r'[A-Za-z0-9_-]{1,80}', name):
            raise ValueError('invalid run ID')
        path = self.root / name
        if path.is_symlink() or path.resolve().parent != self.root:
            raise ValueError('run outside report root')
        return path

    def runs(self):
        runs = []
        if not self.root.exists():
            return runs
        for path in sorted(self.root.iterdir(), reverse=True):
            if len(runs) >= 100:
                break
            if not path.is_dir() or path.is_symlink():
                continue
            try:
                item = read_json(self.directory(path.name) / 'summary.json')
                runs.append({'id': path.name, 'status': item.get('status'),
                             'profile': item.get('audit_profile'), 'started_at': item.get('started_at')})
            except (OSError, ValueError):
                continue
        return runs

    def run(self, name):
        directory = self.directory(name)
        summary = read_json(directory / 'summary.json')
        try:
            rows = [project_sample(r) for r in tail_rows(directory / 'metrics.ndjson')]
        except FileNotFoundError:
            rows = []
        age = None
        if rows and rows[-1].get('time'):
            try:
                age = time.time() - datetime.fromisoformat(rows[-1]['time'].replace('Z', '+00:00')).timestamp()
            except (ValueError, TypeError):
                pass
        current = summary.get('current') or {}
        context = summary.get('context') or {}
        results = []
        for r in summary.get('results', [])[-32:]:
            m = r.get('metrics') or {}
            results.append({'scenario': r.get('scenario'), 'phase': r.get('phase'),
                            'repetition': r.get('repetition'), 'kind': r.get('kind'),
                            'metrics': {k: finite(m.get(k)) for k in (
                                'received_bps', 'requests_per_second', 'errors', 'loss_pct',
                                'jitter_ms')},
                            'latency_bucket_upper_ms': {k: finite((m.get('latency_ms_bucket_upper_bounds') or {}).get(k))
                                                        for k in ('p50', 'p95', 'p99')}})
        cfg = summary.get('config') or {}
        return {'id': name, 'status': summary.get('status'), 'profile': summary.get('audit_profile'),
                'thresholds': {k: (summary.get('measurement_config') or {}).get(k)
                               for k in ('single_core_pct', 'temperature_start_c', 'temperature_stop_c')},
                'started_at': summary.get('started_at'), 'finished_at': summary.get('finished_at'),
                'context': {k: context.get(k) for k in ('campaign_id', 'test_id', 'comparison_id', 'workload_id')},
                'current': {k: current.get(k) for k in ('scenario', 'repetition', 'phase', 'phase_started_at', 'phase_duration_seconds')},
                'offered': {k: (current.get('offered') or {}).get(k) for k in ('kind', 'mbps', 'rate', 'concurrency', 'parallel')},
                'reason': str(summary.get('reason') or '')[:512],
                'gaps': [str(gap)[:256] for gap in summary.get('gaps', [])][:32],
                'results': results, 'samples': rows, 'sample_age_seconds': age,
                'stale': age is None or age < -5 or age > max(20, 3 * cfg.get('sample_interval', 5)),
                'generated_at': datetime.now(timezone.utc).isoformat(),
                'boundary': 'Observation only. Hypothesis verdict requires paired comparisons and audit acceptance.'}


def handler(store):
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *_args):
            pass

        def do_GET(self):
            # Literal loopback host + same-origin browser requests prevent DNS rebinding.
            host = self.headers.get('Host', '')
            if not re.fullmatch(r'(?:127\.0\.0\.1|localhost)(?::[0-9]+)?', host):
                return self.send_error(403)
            if self.headers.get('Sec-Fetch-Site') not in (None, 'same-origin', 'none'):
                return self.send_error(403)
            origin = self.headers.get('Origin')
            if origin and origin != 'http://' + host:
                return self.send_error(403)
            route = urlsplit(self.path)
            try:
                if route.query or route.fragment:
                    return self.send_error(400)
                if route.path == '/api/runs':
                    body = json.dumps(store.runs(), allow_nan=False).encode()
                    mime = 'application/json'
                elif route.path.startswith('/api/run/'):
                    body = json.dumps(store.run(route.path[9:]), allow_nan=False).encode()
                    mime = 'application/json'
                elif route.path in ('/', '/monitor.js', '/monitor.css'):
                    filename = 'index.html' if route.path == '/' else route.path[1:]
                    body = (ASSETS / filename).read_bytes()
                    mime = {'index.html': 'text/html', 'monitor.js': 'text/javascript', 'monitor.css': 'text/css'}[filename]
                else:
                    return self.send_error(404)
            except (OSError, ValueError, TypeError):
                return self.send_error(404, 'Run unavailable or incomplete')
            self.send_response(200)
            self.send_header('Content-Type', mime + '; charset=utf-8')
            self.send_header('Content-Length', str(len(body)))
            self.send_header('Cache-Control', 'no-store')
            self.send_header('X-Content-Type-Options', 'nosniff')
            self.send_header('Content-Security-Policy', "default-src 'self'; object-src 'none'; frame-ancestors 'none'; base-uri 'none'")
            self.end_headers()
            self.wfile.write(body)
    return Handler


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--reports', type=Path, required=True)
    parser.add_argument('--port', type=int, default=8787)
    args = parser.parse_args()
    if not 1024 <= args.port <= 65535:
        parser.error('port must be 1024..65535')
    server = ThreadingHTTPServer(('127.0.0.1', args.port), handler(Store(args.reports)))
    server.daemon_threads = True
    print(f'Read-only test monitor: http://127.0.0.1:{args.port}/', flush=True)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        server.server_close()


if __name__ == '__main__':
    main()

File: roles/ngfw_lab/files/monitor/monitor.js
'use strict';
const $ = id => document.getElementById(id);
const fmt = (v, unit='', digits=1) => typeof v === 'number' && Number.isFinite(v) ? `${v.toLocaleString('ru-RU',{maximumFractionDigits:digits})}${unit}` : '—';
const phases = {warmup:'Разогрев',measure:'Измерение',idle:'Восстановление'};
const states = {running:'Идёт прогон',completed:'Прогон завершён',aborted:'Остановлен'};
const make = (tag, text, cls) => {const e=document.createElement(tag);if(text!==undefined)e.textContent=text;if(cls)e.className=cls;return e;};
function badge(el,text,cls){el.textContent=text;el.className=`badge ${cls||''}`;}
function chart(id,rows,series,fixedMax){
 const svg=$(id);svg.replaceChildren();const ns='http://www.w3.org/2000/svg';
 const node=(tag,attrs,text)=>{const n=document.createElementNS(ns,tag);for(const[k,v]of Object.entries(attrs))n.setAttribute(k,v);if(text!==undefined)n.textContent=text;svg.append(n);return n;};
 const vals=series.flatMap(s=>rows.map(s.get)).filter(v=>typeof v==='number'&&Number.isFinite(v));
 if(!vals.length){node('text',{x:280,y:80,'text-anchor':'middle',fill:'#8491a3','font-size':13},'Нет измерений');return;}
 const times=rows.map(r=>Date.parse(r.time));const first=times[0],span=times.at(-1)-first;
 const max=fixedMax||Math.max(...vals,1)*1.1, x=i=>44+(span>0?(times[i]-first)/span:0)*506, y=v=>132-116*v/max;
 for(let i=0;i<3;i++){const value=max*i/2;node('line',{x1:44,x2:550,y1:y(value),y2:y(value),stroke:'#e5ebf2'});node('text',{x:36,y:y(value)+4,'text-anchor':'end',fill:'#7b8a9e','font-size':10},fmt(value));}
 for(const s of series){let d='',pen=false;rows.forEach((r,i)=>{const v=s.get(r);if(typeof v!=='number'||!Number.isFinite(v)){pen=false;return;}d+=`${pen?'L':'M'}${x(i)},${y(v)} `;pen=true;});node('path',{d,fill:'none',stroke:s.color,'stroke-width':2});}
 for(const[i,anchor]of [[0,'start'],[rows.length-1,'end']]){const d=new Date(rows[i]?.time);node('text',{x:x(i),y:153,'text-anchor':anchor,fill:'#7b8a9e','font-size':10},Number.isNaN(d.getTime())?'':d.toLocaleTimeString('ru-RU'));}
}
function render(data){
 $('empty').hidden=true;$('content').hidden=false;
 const c=data.current||{}, last=data.samples.at(-1)||{}, ctx=data.context||{};
 $('test').textContent=[ctx.test_id,ctx.campaign_id,`Профиль ${data.profile||'—'}`].filter(Boolean).join(' / ');
 $('scenario').textContent=c.scenario||'Подготовка';
 let elapsed=Number.isFinite(Date.parse(c.phase_started_at))?(Date.now()-Date.parse(c.phase_started_at))/1000:null;
 const remaining=data.status==='running'&&elapsed!==null&&c.phase_duration_seconds ? Math.max(0,c.phase_duration_seconds-elapsed):null;
 $('phase').textContent=`${phases[c.phase]||'Ожидание'} · повтор ${c.repetition||'—'}${remaining!==null?' · осталось '+Math.ceil(remaining)+' с':''}`;
 badge($('status'),states[data.status]||data.status,data.status==='aborted'?'bad':data.status==='running'?'good':'');
 badge($('connection'),data.status!=='running'?'Архивный прогон':data.stale?'Данные устарели':'На связи',data.stale&&data.status==='running'?'bad':data.status==='running'?'good':'');
 $('freshness').textContent=`Последний отсчёт: ${fmt(data.sample_age_seconds,' с назад',0)}`;
 const alarms=[];if(data.reason)alarms.push(data.reason);if(data.stale&&data.status==='running')alarms.push('Свежих измерений нет. Это не подтверждение безопасного состояния.');
 if(last.lost>0)alarms.push(`Потери аудита: ${last.lost}`);
 $('alarm').hidden=!alarms.length;$('alarm').textContent=alarms.join(' · ');
 $('throughput').textContent=fmt(last.endpoint_rx_bps===null?null:last.endpoint_rx_bps/1e6,' Мбит/с');
 $('cpu').textContent=fmt(last.guest_cpu_pct,' %');
 const cores=Object.entries(last.cores||{}).filter(([,v])=>v!==null).sort((a,b)=>Number(a[0].slice(3))-Number(b[0].slice(3)));
 $('hottest').textContent='Самое занятое ядро: '+fmt(cores.length?Math.max(...cores.map(x=>x[1])):null,' %');
 const temps=Object.values(last.temperature_c||{}).filter(x=>x!==null);
 $('temperature').textContent=fmt(temps.length?Math.max(...temps):null,' °C');
 $('backlog').textContent=fmt(last.backlog_pct,' %');$('lost').textContent='Потерянные записи: '+fmt(last.lost,'',0);
 const o=data.offered||{};$('offered').textContent='Задано: '+(o.mbps?fmt(o.mbps,' Мбит/с'):o.rate?fmt(o.rate,' операций/с'):'—');
 $('lograte').textContent='audit.log: '+fmt(last.audit_bytes_s===null?null:last.audit_bytes_s/1024,' КиБ/с')+(last.rotation?' · ротация, скорость в этом интервале неизвестна':'');
 chart('traffic-chart',data.samples,[{get:r=>r.endpoint_rx_bps===null?null:r.endpoint_rx_bps/1e6,color:'#3679e7'}]);
 chart('cpu-chart',data.samples,[{get:r=>r.guest_cpu_pct,color:'#3679e7'},{get:r=>r.host_cpu_pct,color:'#93a0b2'}],100);
 chart('audit-chart',data.samples,[{get:r=>r.backlog_pct,color:'#9774cd'}],100);
 chart('probe-chart',data.samples,[{get:r=>r.service_probe_ms,color:'#379984'}]);
 $('cores').replaceChildren(...cores.map(([name,v])=>{const e=make('div',name,'core'+(typeof data.thresholds.single_core_pct==='number'&&v>=data.thresholds.single_core_pct?' hot':''));e.append(make('strong',fmt(v,'%')));return e;}));
 if(!cores.length)$('cores').append(make('p','Данные по ядрам не собраны'));
 const disk=Object.entries(last.disks||{}).map(([name,v])=>make('p',`${name}: await ${fmt(v.await_ms,' мс')} · запись ${fmt(v.write_bytes_s===null?null:v.write_bytes_s/1024,' КиБ/с')} · очередь ${fmt(v.avg_queue_depth)}`));
 disk.push(make('p',Object.entries(last.guest_cpu_breakdown||{}).map(([k,v])=>k+': '+fmt(v,'%')).join(' · ')));
 for(const p of last.processes||[])disk.push(make('p',`${p.comm}: ${fmt(p.cpu_one_core_pct,'% одного ядра')} · ${p.state} · ожидание ${p.wchan||'—'}`));
 disk.push(make('p','Изменение счётчика audit wait: '+fmt(last.audit_wait_delta)+' (единицы установленного ядра)'));
 const forwarder=last.forwarder||{};
 disk.push(make('p','Счётчики отправителя: '+(forwarder.age_seconds===undefined?'не собраны':fmt(forwarder.age_seconds,' с назад'))));
 for(const[name,v]of Object.entries(forwarder.counters||{}))disk.push(make('p',`${name}: очередь ${fmt(v.size,'',0)} · full ${fmt(v.full,'',0)} · dropped ${fmt(v['discarded.full'],'',0)}`));
 for(const v of last.collector?.counters||[])disk.push(make('p',`Коллектор ${v.id}: ${v.counter} ${fmt(v.value,'',0)}`));
 $('disk').replaceChildren(...disk);
 $('results').replaceChildren(...data.results.filter(r=>r.phase==='measure').map(r=>{const tr=make('tr');const m=r.metrics;[r.scenario,r.repetition,m.received_bps!==null?fmt(m.received_bps/1e6,' Мбит/с'):fmt(m.requests_per_second,' оп/с'),m.loss_pct!==null?fmt(m.loss_pct,' %'):fmt(m.errors,' ошибок',0),fmt(r.latency_bucket_upper_ms.p95)].forEach(v=>tr.append(make('td',v)));return tr;}));
 $('gaps').replaceChildren(...[...data.gaps,...(last.unavailable||[]).map(v=>'Недоступен источник: '+v)].map(x=>make('li',x)));
}
let busy=false;
async function refresh(){
 if(busy)return;busy=true;
 try{const res=await fetch('/api/runs',{cache:'no-store'});if(!res.ok)throw Error('HTTP '+res.status);const runs=await res.json();
 const selected=$('runs').value; $('runs').replaceChildren(...runs.map(r=>{const o=make('option',`${r.id} · ${r.profile||'—'}`);o.value=r.id;return o;}));
 if(runs.some(r=>r.id===selected))$('runs').value=selected;
 if(!runs.length){$('empty').hidden=false;$('content').hidden=true;badge($('connection'),'Нет прогонов','');return;}
 const detail=await fetch('/api/run/'+encodeURIComponent($('runs').value),{cache:'no-store'});if(!detail.ok)throw Error('Run unavailable');render(await detail.json());
 }catch(error){badge($('connection'),'Нет связи','bad');$('alarm').hidden=false;$('alarm').textContent='Обновление недоступно. Показанные значения могут быть устаревшими.';}
 finally{busy=false;}
}
$('runs').addEventListener('change',refresh);refresh();setInterval(refresh,2000);
